### PR TITLE
allow the WebAudio backend to accept prerendered peaks just like the MediaElement backend

### PR DESCRIPTION
**Pull request by cdelguercio**
I added the ability to pass peaks from the wavesurfer.load() method to the WebAudio media backend. I haven't done much real testing on anything other than the site that I'm currently using wavesurfer on.

### CODE REVIEW

**Comment by katspaugh** [comment]
Thanks for the contribution, but what is the point of passing the peaks if the audio is decoded anyway?

**Comment by cdelguercio** [comment]
Performance, it take around 5ms to run getPeaks on my (fast) laptop. If you have 20+ songs on a page and you're on a phone the time saving could easily reach a few hundred ms. I'm prerendering the peaks serverside and then passing them to the client with the songs. The peaks themselves are so small that the added network time is easily sub-millisecond.

**Comment by thijstriemstra** [comment]
@cdelguercio there are some build errors..

File: dist/wavesurfer.js [hunk 1]
```diff
@@ -847,6 +851,10 @@ WaveSurfer.WebAudio = {
      * of peaks consisting of (max, min) values for each subrange.
      */
     getPeaks: function (length) {
+        if( this.peaks ) {
```
Comment (katspaugh) [inline dist/wavesurfer.js]: Please change to `if (this.peaks) {`.

File: dist/wavesurfer.js [hunk 2]
```diff
@@ -333,18 +333,13 @@ var WaveSurfer = {
...
 
-    /**
-     * Loads audio using Web Audio buffer backend.
-     */
-    loadBuffer: function (url) {
-        this.empty();
-        // load via XHR and render all at once
-        return this.getArrayBuffer(url, this.loadArrayBuffer.bind(this));
+    loadWebAudio: function(url, peaks) {
+        return this.loadBuffer(url, peaks);
```
Comment (katspaugh) [inline dist/wavesurfer.js]: What is the point of this function if it just calls another function with the same signature?

File: src/webaudio.js [hunk 3]
```diff
@@ -184,6 +184,10 @@ WaveSurfer.WebAudio = {
      * of peaks consisting of (max, min) values for each subrange.
      */
     getPeaks: function (length) {
+        if( this.peaks ) {
```
Comment (katspaugh) [inline src/webaudio.js]: `if (this.peaks) {`

File: dist/wavesurfer.js (hunk 2 above)
Comment (thijstriemstra) [inline dist/wavesurfer.js]: was thinking the same!

File: src/wavesurfer.js [hunk 4]
```diff
@@ -283,17 +283,17 @@ var WaveSurfer = {
     /**
      * Internal method.
      */
-    loadArrayBuffer: function (arraybuffer) {
+    loadArrayBuffer: function (arraybuffer, peaks) {
         this.decodeArrayBuffer(arraybuffer, function (data) {
-            this.loadDecodedBuffer(data);
+            this.loadDecodedBuffer(data, peaks);
```
Comment (katspaugh) [inline src/wavesurfer.js]: I wonder if we could make it work in the same way as Media Element. I.e. draw the waveform first, and then decode the buffer.

**Comment by cdelguercio** [comment]
@katspaugh you read my mind. I'm actually working on a PR right now where we allow the user to JUST load the peaks, and then load the audio afterwards. This would copy what soundcloud does where the audio isn't loaded until the user actually hits play. The benefit would be a massive decrease in load time of pages that have many songs on them. I plan to submit a new PR later today.

Regarding the extra function call: I put that in for readability. When I first read through the code there it was a little confusing because the 'loadMediaElement' case called this.loadMediaElement but the 'webAudio' case called this.loadBuffer. (loadBuffer has a very generic meaning until you understand that buffer is the audio info that if owned by the audio backend) I'm from the school of thought where it's okay to write a little bit of extra code in order to make the semantics work out better. I can change this if it bothers everyone.

**Comment by katspaugh** [comment]
Awesome! Then I'll wait for the next PR.

Regarding the function name, we could rename the `loadBuffer` to `loadWebAudio`. I don't think it was called directly from app code.

**Comment by thijstriemstra** [comment]
-1 on renaming `loadBuffer`, it'll break existing code out there.

**Comment by katspaugh** [comment]
Closing in favour of #733 or #734.

**Pull request closed without merging**